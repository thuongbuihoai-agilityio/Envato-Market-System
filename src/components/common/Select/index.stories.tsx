import type { Meta, StoryObj } from '@storybook/react';

// Components
import { Select } from '@components/index';

// Mocks
import { OPTIONS } from '@mocks/select';

const meta: Meta<typeof Select> = {
  title: 'Custom Components/Select',
  component: Select,
};

export default meta;
type Story = StoryObj<typeof Select>;

export const Primary: Story = {
  args: {
    options: OPTIONS,
  },
};

export const Secondary: Story = {
  args: {
    options: OPTIONS,
    variant: 'secondary',
  },
};

export const NoBorder: Story = {
  args: {
    options: OPTIONS,
    variant: 'no-border',
  },
};

export const NoBackground: Story = {
  args: {
    options: OPTIONS,
    variant: 'no-background',
  },
};
