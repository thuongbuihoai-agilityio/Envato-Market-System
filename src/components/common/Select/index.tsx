import {
  Button,
  Menu,
  MenuButton,
  MenuItem,
  MenuList,
  Text,
} from '@chakra-ui/react';
import { MouseEventHandler, memo, useState } from 'react';
import areEqual from 'react-fast-compare';

const PRIMARY_BG_COLOR = 'background.component.select.primary';
const SECONDARY_BG_COLOR = 'background.component.select.primary';
const NO_COLOR = 'background.component.select.secondary';
const NO_BG = 'background.component.select.noBackground';

const BORDER_STYLES = {
  borderWidth: '1px',
  borderStyle: 'solid',
  borderColor: 'border.primary',
};
const VARIANTS = {
  primary: {
    bgColor: SECONDARY_BG_COLOR,
    ...BORDER_STYLES,
    _hover: {
      background: SECONDARY_BG_COLOR,
      ...BORDER_STYLES,
    },
    _active: {
      background: SECONDARY_BG_COLOR,
    },
    _dark: {
      border: 'none',
    },
  },
  secondary: {
    bgColor: NO_COLOR,
    ...BORDER_STYLES,
    _hover: {
      background: NO_COLOR,
      ...BORDER_STYLES,
    },
    _active: {
      background: NO_COLOR,
    },
  },
  'no-border': {
    bgColor: PRIMARY_BG_COLOR,

    _hover: {
      background: PRIMARY_BG_COLOR,
    },
    _active: {
      background: PRIMARY_BG_COLOR,
    },
  },
  'no-background': {
    bgColor: NO_BG,

    _hover: {
      background: NO_BG,
    },
    _active: {
      background: NO_BG,
    },
  },
};
const SIZES = {
  sm: {
    px: 3,
  },
  md: {
    px: 5,
  },
};

export type TOption = {
  value: string;
  label: string;
};

type TSelectProps = {
  title?: string;
  options?: TOption[];
  size?: keyof typeof SIZES;
  variant?: keyof typeof VARIANTS;
  renderTitle?: (option: TOption) => JSX.Element;
  renderOption?: (option: TOption) => JSX.Element;
  onSelect?: (option: TOption) => void;
};

const SelectComponent = ({
  options = [],
  variant = 'primary',
  size = 'md',
  renderTitle,
  renderOption,
  onSelect,
}: TSelectProps): JSX.Element => {
  const [selected, setSelected] = useState<number>(0);

  return (
    <Menu matchWidth>
      <MenuButton
        as={Button}
        h="100%"
        textAlign="left"
        borderRadius="lg"
        {...VARIANTS[variant]}
        {...SIZES[size]}
      >
        {renderTitle && renderTitle(options[selected])}
        {!renderTitle && (
          <Text as="span" fontSize="sm" textTransform="capitalize">
            {options[selected]?.label}
          </Text>
        )}
      </MenuButton>
      <MenuList
        p={0}
        minW={0}
        boxShadow="lg"
        borderRight="lg"
        border="none"
        overflow="hidden"
        bg="background.component.selectList"
      >
        {options.map((option: TOption, index: number): JSX.Element => {
          const { label } = option;
          const handleSelected: MouseEventHandler = () => {
            setSelected(index);
            onSelect && onSelect(option);
          };

          return (
            <MenuItem
              key={label}
              bg="background.component.selectList"
              borderRadius="none"
              py={2}
              _hover={{
                borderColor: 'transparent',
                _light: { bgColor: 'secondary.200' },
                _dark: { bgColor: 'secondary.400' },
              }}
              _focus={{ borderColor: 'transparent', outline: 'none' }}
              onClick={handleSelected}
              {...SIZES[size]}
              type="submit"
            >
              <Text
                fontSize="sm"
                fontWeight="semibold"
                textTransform="capitalize"
              >
                {renderOption ? renderOption(option) : label}
              </Text>
            </MenuItem>
          );
        })}
      </MenuList>
    </Menu>
  );
};

const Select = memo(SelectComponent, areEqual);

export default Select;
