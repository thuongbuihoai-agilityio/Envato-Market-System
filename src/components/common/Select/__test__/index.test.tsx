import { fireEvent, render } from '@testing-library/react';

// Components
import Select from '..';

// Mocks
import { OPTIONS } from '@mocks/index';

describe('Select', () => {
  it('Match snapshot', () => {
    const { container } = render(<Select options={OPTIONS} />);

    expect(container).toMatchSnapshot();
  });

  it('Open list item', () => {
    const { getByRole, container } = render(<Select options={OPTIONS} />);
    const button = getByRole('button');

    fireEvent.click(button);
    expect(container.querySelectorAll('button').length).toBe(
      OPTIONS.length + 1,
    );
  });

  it('Select one item', () => {
    const { getByRole, container } = render(<Select options={OPTIONS} />);
    const button = getByRole('button');

    fireEvent.click(button);

    const buttons = container.querySelectorAll('button');

    expect(buttons.length).toBe(OPTIONS.length + 1);
    fireEvent.click(buttons[2]);
    expect(getByRole('button').textContent).toBe(OPTIONS[1].label);
  });
});
