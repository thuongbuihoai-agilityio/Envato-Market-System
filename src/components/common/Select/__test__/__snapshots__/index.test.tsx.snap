// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Match snapshot 1`] = `
<div>
  <button
    aria-controls="menu-list-:r1:"
    aria-expanded="false"
    aria-haspopup="menu"
    class="chakra-button chakra-menu__menu-button css-lyu4i6"
    id="menu-button-:r1:"
    type="button"
  >
    <span
      class="css-xl71ch"
    >
      <span
        class="chakra-text css-cwazn2"
      >
        github
      </span>
    </span>
  </button>
  <div
    class="css-0"
    style="visibility: hidden; position: absolute; inset: 0 auto auto 0;"
  >
    <div
      aria-orientation="vertical"
      class="chakra-menu__menu-list css-ta0hp9"
      id="menu-list-:r1:"
      role="menu"
      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
      tabindex="-1"
    >
      <button
        class="chakra-menu__menuitem css-1oz22gy"
        data-index="0"
        id="menu-list-:r1:-menuitem-:r2:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <p
          class="chakra-text css-1198d6z"
        >
          github
        </p>
      </button>
      <button
        class="chakra-menu__menuitem css-1oz22gy"
        data-index="1"
        id="menu-list-:r1:-menuitem-:r3:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <p
          class="chakra-text css-1198d6z"
        >
          facebook
        </p>
      </button>
      <button
        class="chakra-menu__menuitem css-1oz22gy"
        data-index="2"
        id="menu-list-:r1:-menuitem-:r4:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <p
          class="chakra-text css-1198d6z"
        >
          instagram
        </p>
      </button>
    </div>
  </div>
</div>
`;

exports[`Select Open list item 1`] = `
<div>
  <button
    aria-controls="menu-list-:r1:"
    aria-expanded="true"
    aria-haspopup="menu"
    class="chakra-button chakra-menu__menu-button css-lyu4i6"
    data-active=""
    id="menu-button-:r1:"
    type="button"
  >
    <span
      class="css-xl71ch"
    >
      <span
        class="chakra-text css-cwazn2"
      >
        github
      </span>
    </span>
  </button>
  <div
    class="css-0"
    data-popper-escaped=""
    data-popper-placement="bottom-start"
    data-popper-reference-hidden=""
    style="visibility: visible; position: absolute; inset: 0 auto auto 0; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 8px); width: 0px; --popper-transform-origin: top left;"
  >
    <div
      aria-orientation="vertical"
      class="chakra-menu__menu-list css-ta0hp9"
      id="menu-list-:r1:"
      role="menu"
      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
      tabindex="-1"
    >
      <button
        class="chakra-menu__menuitem css-1oz22gy"
        data-index="0"
        id="menu-list-:r1:-menuitem-:r2:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <p
          class="chakra-text css-1198d6z"
        >
          github
        </p>
      </button>
      <button
        class="chakra-menu__menuitem css-1oz22gy"
        data-index="1"
        id="menu-list-:r1:-menuitem-:r3:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <p
          class="chakra-text css-1198d6z"
        >
          facebook
        </p>
      </button>
      <button
        class="chakra-menu__menuitem css-1oz22gy"
        data-index="2"
        id="menu-list-:r1:-menuitem-:r4:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <p
          class="chakra-text css-1198d6z"
        >
          instagram
        </p>
      </button>
    </div>
  </div>
</div>
`;
